microceph/api/remote.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/canonical/microceph/microceph/interfaces"
+	"github.com/canonical/microcluster/rest"
+	"github.com/canonical/microcluster/state"
+	"github.com/gorilla/mux"
+)
+
+var remoteCmd = rest.Endpoint{
+	Path: "client/remotes",
+	Get:  rest.EndpointAction{Handler: CmdRemoteGet, ProxyTarget: false},
+}
+
+var remoteNameCmd = rest.Endpoint{
+	Path:   "client/remotes/{name}",
+	Put:    rest.EndpointAction{Handler: CmdRemotePut, ProxyTarget: false},
+	Get:    rest.EndpointAction{Handler: CmdRemoteGet, ProxyTarget: false},
+	Delete: rest.EndpointAction{Handler: CmdRemoteDelete, ProxyTarget: false},
+}
+
+var CmdRemotePut = func(state *state.State, r *http.Request) response.Response {
+	var req types.Remote
+
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	err = RenderConfAndKeyringFiles(req.Name, req.LocalName, req.Config)
+	if err != nil {
+		return response.InternalError(fmt.Errorf("couldn't render files: %w", err))
+	}
+
+	if !req.RenderOnly {
+		// Asynchronously persist this on db and send request to other cluster members.
+		go func() {
+			err := PersisteRemoteAndConfigs(interfaces.CephState{State: state}, req)
+			if err != nil {
+				logger.Errorf("failed to persiste remote: %s", err.Error())
+			}
+
+			// Send render only request to remaining cluster members.
+			req.RenderOnly = true
+			err = client.SendRemoteImportToClusterMembers(state, req)
+			if err != nil {
+				logger.Errorf("failed to forward request to cluster: %s", err.Error())
+			}
+		}()
+	}
+
+	return response.EmptySyncResponse
+}
+
+var CmdRemoteGet = func(state *state.State, r *http.Request) response.Response {
+	// PathUnescape will NOT fail if no name is provided in API request.
+	// Additionally, remoteName in that case is initialised to "".
+	remoteName, err := url.PathUnescape(mux.Vars(r)["name"])
+	if err != nil {
+		logger.Error(err.Error())
+		return response.InternalError(err)
+	}
+
+	remotes, err := database.GetRemoteDb(*state, remoteName)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if len(remotes) == 0 {
+		return response.SmartError(fmt.Errorf("no remotes configured"))
+	}
+
+	return response.SyncResponse(true, remotes)
+}
+
+var CmdRemoteDelete = func(state *state.State, r *http.Request) response.Response {
+	remoteName, err := url.PathUnescape(mux.Vars(r)["name"])
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	// Note(utkarshbhatthere): TODO for when remote replication is implemented.
+	// [ ] add check for remote replication before deleting remotes.
+	err = database.DeleteRemoteDb(*state, remoteName)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	return response.EmptySyncResponse
+}
+
+/*****************HELPER FUNCTIONS**************************/
+// PersisteRemoteAndConfigs adds the remote record to dqlite.
+var PersisteRemoteAndConfigs = func(s interfaces.StateInterface, remote types.Remote) error {
```

Comment:
Naming nit: suggest s/PersisteRemoteAndConfigs/PersistRemoteAndConfigs/

Also, is there a reason this is a func variable and not a regular function? And, does it need to be public?

Wrt to organization, I'd lean towards putting DB-accessing funcs into database/remote_extras.go 

Same for the below functions

Reply:
Ack